src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivitiesService.kt

```diff
@@ -299,6 +299,17 @@ class ActivitiesService(
       .filter { old -> !old.hasFutureStartDate() } // ignore future rates not included in updates - so they are deleted
       .filter { old -> !newPayRates.containsRate(old) }
       .map { old -> old.expire() }
+      .also { expiredPayRates -> expiredPayRates.throwIfPayBandsInUse() }
+
+  private fun List<CourseActivityPayRate>.throwIfPayBandsInUse() =
+    this.forEach { activityPayRate ->
+      offenderProgramProfileRepository.findByCourseActivity(activityPayRate.id.courseActivity)
+        .filter { profile -> profile.isUsingPayBand(activityPayRate.payBand.code) }
+        .map { offender -> offender.offenderBooking.offender.nomsId }
+        .toList()
+        .takeIf { nomsIds -> nomsIds.isNotEmpty() }
+        ?.run { throw BadDataException("Pay band ${activityPayRate.payBand.code} is allocated to offender(s) $this") }
```

Comment:
Feels like this can be shorter, not sure why you need the map for instance. Feels like it can be a "find" or something